dbos/system_database.py

```diff
@@ -1030,22 +1034,58 @@ def enqueue(self, workflow_id: str, queue_name: str) -> None:
                 .on_conflict_do_nothing()
             )
 
-    def start_queued_workflows(
-        self, queue_name: str, concurrency: Optional[int]
-    ) -> List[str]:
+    def start_queued_workflows(self, queue: "Queue") -> List[str]:
+        start_time_ms = int(time.time() * 1000)
+        if queue.limiter is not None:
+            limiter_duration_ms = int(queue.limiter["duration"] * 1000)
         with self.engine.begin() as c:
-            query = sa.select(SystemSchema.job_queue.c.workflow_uuid).where(
-                SystemSchema.job_queue.c.queue_name == queue_name
+            # Execute with snapshot isolation to ensure multiple workers respect limits
+            c.execute(sa.text("SET TRANSACTION ISOLATION LEVEL REPEATABLE READ"))
+
+            # If there is a limiter, compute how many functions have started in its duration.
+            if queue.limiter is not None:
+                query = (
+                    sa.select(sa.func.count())
+                    .select_from(SystemSchema.job_queue)
+                    .where(SystemSchema.job_queue.c.started_at_epoch_ms.isnot(None))
+                    .where(
+                        SystemSchema.job_queue.c.started_at_epoch_ms
+                        > start_time_ms - limiter_duration_ms
+                    )
+                )
+                num_recent_queries = c.execute(query).fetchone()[0]  # type: ignore
```

Comment:
I think if `num_recent_queries >= queue.limiter["max"]` we should directly return. No need to have another query.